AK/URLParser.cpp

```diff
@@ -0,0 +1,666 @@
+            // NOTE: Verify that the assumptions required for this simplification are correct.
+            VERIFY(url.paths().size() == 1 && url.paths()[0].is_empty());
+            if (code_point == '?') {
+                url.m_paths[0] = buffer.to_string();
+                url.set_query("");
+                state = State::Query;
+            } else if (code_point == '#') {
+                url.m_paths[0] = buffer.to_string();
+                url.set_fragment("");
+                state = State::Fragment;
+            } else {
+                if (code_point != 0 && !is_url_code_point(code_point) && code_point != '%')
+                    report_validation_error();
+                // FIXME: If c is U+0025 (%) and remaining does not start with two ASCII hex digits, validation error.
+                if (code_point != 0) {
+                    buffer.append(URL::percent_encode(code_point, URL::PercentEncodeSet::C0Control));
+                } else {
+                    url.m_paths[0] = buffer.to_string();
+                }
+            }
+            break;
+        case State::Query:
+            if (code_point == '#' || code_point == 0) {
+                VERIFY(url.query() == "");
+                auto query_percent_encode_set = url.is_special() ? URL::PercentEncodeSet::SpecialQuery : URL::PercentEncodeSet::Query;
+                url.set_query(URL::percent_encode(buffer.to_string(), query_percent_encode_set));
+                buffer.clear();
+                if (code_point == '#') {
+                    url.set_fragment("");
+                    state = State::Fragment;
+                }
+            } else if (code_point != 0) {
+                if (!is_url_code_point(code_point) && code_point != '%')
+                    report_validation_error();
+                // FIXME: If c is U+0025 (%) and remaining does not start with two ASCII hex digits, validation error.
+                buffer.append_code_point(code_point);
+            }
+            break;
+        case State::Fragment:
+            // NOTE: This does not follow the spec exactly but rather uses the buffer and only sets the fragment on EOF.
+            if (code_point) {
+                if (!is_url_code_point(code_point) && code_point != '%')
+                    report_validation_error();
+                // FIXME: If c is U+0025 (%) and remaining does not start with two ASCII hex digits, validation error.
+                buffer.append_code_point(code_point);
+            } else {
+                url.set_fragment(buffer.to_string());
+                buffer.clear();
+            }
+            break;
+        default:
+            VERIFY_NOT_REACHED();
+        }
+
+        if (iterator.done())
+            break;
+        ++iterator;
+    }
+
+    // NOTE: Percent-decode all fields that can contain encoded codepoints and aren't already decoded by the parser algorithm.
+    //       The host is already decoded by parse_host().
```

Comment:
I didn't do something like this, because this "problem" should by internal to the `URLParser` class. After the URL is returned for the `URLParser`, it should always contain decoded data, unless you call `serialize()` on it.

But thinking about this just helped me find a bug: If e.g. the URL query is set from the base URL, it is already decoded. We can fix this by percent decoding the data when setting the members inside the parser algorithm.

Reply:
I have now moved the percent decoding to the places where the members are set inside the parser algorithm. I will add a comment at the top of the algorithm explaining this.